prefork/prefork.go

```diff
@@ -110,30 +120,84 @@ func (p *Prefork) setTCPListenerFiles(addr string) error {
 	return nil
 }
 
-func (p *Prefork) prefork(addr string) error {
-	chErr := make(chan error, 1)
-
+func (p *Prefork) prefork(addr string) (err error) {
 	if !p.Reuseport {
-		if err := p.setTCPListenerFiles(addr); err != nil {
-			return err
+		if err = p.setTCPListenerFiles(addr); err != nil {
+			return
 		}
 
-		defer p.ln.Close()
+		defer func() {
+			if err == nil {
+				err = p.ln.Close()
+			}
+		}()
+	}
+
+	type procSig struct {
+		pid int
+		err error
 	}
 
-	for i := 0; i < runtime.GOMAXPROCS(0); i++ {
+	goMaxProcs := runtime.GOMAXPROCS(0)
+	sigCh := make(chan procSig, goMaxProcs)
+	childProcs := make(map[int]*exec.Cmd)
+
+	defer func() {
+		for _, proc := range childProcs {
+			_ = proc.Process.Kill()
+		}
+	}()
+
+	for i := 0; i < goMaxProcs; i++ {
 		/* #nosec G204 */
 		cmd := exec.Command(os.Args[0], append(os.Args[1:], preforkChildFlag)...)
 		cmd.Stdout = os.Stdout
 		cmd.Stderr = os.Stderr
 		cmd.ExtraFiles = p.files
+		if err = cmd.Start(); err != nil {
+			log.Printf("failed to start a child prefork process, error: %v\n", err)
+			return
+		}
 
+		childProcs[cmd.Process.Pid] = cmd
 		go func() {
-			chErr <- cmd.Run()
+			sigCh <- procSig{cmd.Process.Pid, cmd.Wait()}
 		}()
 	}
 
-	return <-chErr
+	var brokenProcs, completeProcs int
+	for sig := range sigCh {
+		if sig.err != nil {
+			delete(childProcs, sig.pid)
+
+			log.Printf("one of the child prefork processes failed to complete, "+
+				"error: %v", sig.err)
+
+			if brokenProcs++; brokenProcs > p.RecoverThreshold {
+				log.Printf("child prefork processes exit too many times, "+
+					"which exceeds the value of RecoverThreshold(%d), "+
+					"exiting the master process.\n", brokenProcs)
+				err = errors.New("exceeding the value of RecoverThreshold")
+				break
+			}
+
+			/* #nosec G204 */
+			cmd := exec.Command(os.Args[0], append(os.Args[1:], preforkChildFlag)...)
+			cmd.Stdout = os.Stdout
+			cmd.Stderr = os.Stderr
+			cmd.ExtraFiles = p.files
+			childProcs[cmd.Process.Pid] = cmd
```

Comment:
Isn't `cmd.Process.Pid` here still the old value as `cmd.Run()` hasn't started yet?

Reply:
My oversight, fixed it.